src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsOperation.cs

```diff
@@ -132,7 +134,125 @@ private IEnumerable<IEdmOperationParameter> ComputeParameters()
                 }
             }
 
-            return parameters;
+            // Handle the out-of-line optinal parameter annotation for parameters.
+            // First, use the above built parameters to create the full target name, for example: NS.TestFunction(Edm.String, Edm.String, Edm.String)
+            // Then, go through each prameters by visiting the out-of-line optional parameter annotation.
+            // If we find at least one of out-of-line optional parameter annotation, we replace it as an optional parameter.
+            // Otherwise, re-use the built parameter.
+            // Be noted: if a parameter has inline and out-of-line optional parameter, the out-of-line will win.
+            string fullName = Namespace + "." + Name;
+            string fullParametersName = ParameterizedTargetName(parameters);
+
+            List<IEdmOperationParameter> newParameters = new List<IEdmOperationParameter>();
+            foreach (var parameter in parameters)
+            {
+                // for example: NS.TestFunction(Edm.String, Edm.String, Edm.String)/OptionalParameter
+                string fullTargetName = fullName + fullParametersName + "/" + parameter.Name;
+
+                // for example: NS.TestFunction/OptionalParameter
+                string targetName = fullName + "/" + parameter.Name;
+
+                string defaultValue;
+
+                if (TryGetOptionalParameterAnnotation(fullTargetName, targetName, out defaultValue))
+                {
+                    CsdlSemanticsOperationParameter csdlSemanticsParameter = (CsdlSemanticsOperationParameter)parameter;
+                    newParameters.Add(new CsdlSemanticsOptionalParameter(this, (CsdlOperationParameter)csdlSemanticsParameter.Element, defaultValue));
+                }
+                else
+                {
+                    newParameters.Add(parameter);
+                }
+            }
+
+            return newParameters;
+        }
+
+        private bool TryGetOptionalParameterAnnotation(string fullTargetName, string targetName, out string defaultValue)
+        {
+            defaultValue = null;
+            bool isOptional = false;
+
+            List<CsdlSemanticsAnnotations> annotations;
+
+            // OData 4.0 applies annotations to all overloads of a function or action.
+            // We still need to support annotating the non-overloaded version.
+            // So, we should probably first check the fullTargetName and, if that doesn't match, check just the function or action name.
+            bool found = Model.OutOfLineAnnotations.TryGetValue(fullTargetName, out annotations) ? true :
+                Model.OutOfLineAnnotations.TryGetValue(targetName, out annotations);

```

Comment:
What if the function has some annotations against the fullTarget (not including the OptionalParameter annotation) and some against the non-overloaded name (including OptionalParameter the optional parameter annotation).  It appears if there are *any* annotations against the fully qualified name then we don't even look at the non-overloaded version.  We should look to see if the OptionalParameter annotation is in the fullTarget and, if not, see if it is in targetName (and add a test for this case).

Reply:
Synced with @Michael Pizzo, the codes here is to search the out of line annotation using the "FullTargetName" and "TargetName" (FullTargetName is the qualified operation name with the parameter type list, while TargetName is the qualified operation name). It's ok here.

---
In reply to: [223550455](https://github.com/OData/odata.net/pull/1267#discussion_r223550455) [](ancestors = 223550455)